File: 0623-add-one-row-to-tree/0623-add-one-row-to-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if(depth == 1) {
            TreeNode newNode = new TreeNode(val);
            newNode.left = root;
            return newNode;
        }

        Queue<TreeNode> pq = new LinkedList<>();
        pq.add(root);
        int level = 0;
        boolean flag = false;

        while(!pq.isEmpty()) {
            int n = pq.size();
            level++;
            if(flag) break;

            for(int i = 0; i<n; i++) {
                TreeNode curr = pq.remove();

                if(level == depth - 1) {
                    flag = true;

                    TreeNode newNode1 = new TreeNode(val);
                    newNode1.left = curr.left;
                    curr.left = newNode1;

                    TreeNode newNode2 = new TreeNode(val);
                    newNode2.right = curr.right;
                    curr.right = newNode2;
                }
                else {
                    if(curr.left != null) pq.add(curr.left);
                    if(curr.right != null) pq.add(curr.right);
                }
            }
        }

        return root;
    }
}

File: 0623-add-one-row-to-tree/0623-add-one-row-to-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def addOneRow(self, root, val, depth):
        if depth == 1:
            new_node = TreeNode(val)
            new_node.left = root
            return new_node

        queue = deque([root])
        level = 0
        flag = False

        while queue:
            level += 1
            n = len(queue)
            if flag:
                break

            for _ in range(n):
                curr = queue.popleft()

                if level == depth - 1:
                    flag = True

                    new_node1 = TreeNode(val)
                    new_node1.left = curr.left
                    curr.left = new_node1

                    new_node2 = TreeNode(val)
                    new_node2.right = curr.right
                    curr.right = new_node2
                else:
                    if curr.left:
                        queue.append(curr.left)
                    if curr.right:
                        queue.append(curr.right)

        return root
                
        